### PR TITLE
Reviewing the algorithm for calculating workshop balances

### PR DESCRIPTION
**Pull request by viktorcsimma**
There was a strange phenomenon with the workshop balances: their sum was more than 49% of the amount received from kkt payments; in theory, it should be between 30% and 40%.

I've rewritten the algorithm calculating the data in the balances table; now it displays less people paying kkt and it seems to be consistent with the income numbers. @kdmnk could you check it out?

### CODE REVIEW

**Comment by kdmnk** [comment]
I see, yet I still do not understand where the difference is coming from.
You changed two things:
- Check if the user paid: Instead of checking if the user is not returned by the hasToPayKKTNetregInSemester scope (which returns all active collegists who don't have either kkt or netreg transaction in the semester), you check if the individual user has any kkt transaction (which is more database calls, but that's another problem). Since active memberships are checked anyway before, this change should not result in any difference (given that everyone who has netreg transaction also has kkt transaction, and the contrary). 
- Use the paid amount instead of the amount in the config. I see the point, but do we want to have a workshop where they get less amount only because they have some people in the student council who pay 0 ft?  My opinion is that the workshop balance should be calculated based solely on the number of members, and not who paid how much. 

How did you check and validate the new numbers?

If we want to fix this in the long term, write test cases. This module is one of the last crucial parts where we don't have tests yet.
Let me know what you think.

**Comment by viktorcsimma** [comment]
For the amount part, I agree with you; I've changed it back now. For the other part, I'm thinking about how to solve that. Now I've left it, and it seems to work; but if I understand well, the problem with those paying 0 Ft persists. (Or might that be the reason for the difference?)

**Comment by viktorcsimma** [comment]
Now I've changed it to nullable; the numbers are still OK, but if I understand right, there aren't any transactions recorded for student council members, so the problem would still persist. Is that correct?

I've actually tested the code on the live site; it was consistent with the real values, but I still don't know the reason. Maybe I'll take a look at `hasToPayKKTNetregInSemester`.

**Comment by kdmnk** [comment]
I insist on creating tests for this. There are multiple examples now in the project. That should also reveal the answer to our questions.

**Comment by viktorcsimma** [comment]
Have you thought of something like this? I would write more, of course.

Actually, after some seeding, this fails for both the old and the new version failed... There, the sum of workshop balances was 10200 in the old version and 6600 in the new version; while the income from kkt was always 24000. So both of them was too little. I will check this out.

File: tests/Unit/EconomicTest.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        echo(config('custom.workshop_balance_resident'));
+        echo("\n");
+        echo(config('custom.workshop_balance_extern'));
+        echo("\n");
+        */
+
+        $diff1 = config('custom.workshop_balance_resident') * $skkt - $sum;
+        $diff2 = config('custom.workshop_balance_extern') * $skkt - $sum;
+        $this->assertTrue($diff1 * $diff2 <= 0);
+    }
```
Comment (kdmnk) [inline tests/Unit/EconomicTest.php]: Aren't all the values 0 here?

You should test if after someone pays, the workshop balance for the user got modified with the correct amount.
Test externs, residents, with single and multiple workshops.

You can keep this test case but this is not so useful on its own.

**Review by kdmnk — CHANGES_REQUESTED** [review]

**Comment by viktorcsimma** [comment]
Now I think I've written something like that:) @kdmnk could you check it out?

Maybe a feature test for a complete HTTP request would also be useful; but that could be done in a separate PR, if you agree.

**Comment by viktorcsimma** [comment]
One more thing: I had to rewrite `isResident()` and `isExtern()` because I got an error that there is a string somewhere where a model object should be. Do you have an idea what could have caused this?

**Comment by viktorcsimma** [comment]
And maybe there could be another case where the test user has multiple workshops.

File: app/Http/Controllers/StudentsCouncil/EconomicController.php [hunk 2]
```diff
@@ -85,41 +85,24 @@ public function payKKTNetreg(Request $request)
         $validator->validate();
 
         $payer = User::findOrFail($request->user_id);
+        $return_values = $payer->payKKTNetreg($request->kkt, $request->netreg);
```
Comment (kdmnk) [inline app/Http/Controllers/StudentsCouncil/EconomicController.php]: [$kkt, $netreg, $new_internet_expire_date] = $payer->payKKTNetreg(...)

File: app/Models/User.php [hunk 3]
```diff
@@ -681,13 +681,10 @@ public function setCollegist($objectName): void
...
-            return $this->roles()
-            ->where('role_id', Role::collegist()->id)
-            ->where('object_id', RoleObject::firstWhere('name', Role::RESIDENT)->id)
-            ->exists();
-        }
-        return $this->hasRole([Role::COLLEGIST => Role::RESIDENT]);
+        return $this->roles()
+        ->where('role_id', Role::collegist()->id)
+        ->where('object_id', RoleObject::firstWhere('name', Role::RESIDENT)->id)
+        ->exists();
```
Comment (kdmnk) [inline app/Models/User.php]: Just comment the previous one out and add a todo. 
We need to remove the first branch when we detach roles from the applicants so it should just be  $this->hasRole(...).

For the bug you see, idk, we should find it asap, but if you can not you can leave this like that.

File: app/Models/User.php [hunk 4]
```diff
@@ -869,25 +863,64 @@ public function sumOfActiveFreePages(): int
...
+     * and the internet access expiry date.
+     * Returns an array with the two transaction objects
+     * and the new expiry date.
+     */
+    public function payKKTNetreg(int $kkt_amount, int $netreg_amount)
+    {
+        // Creating transactions
+        $kkt = Transaction::create([
+            'checkout_id' => Checkout::studentsCouncil()->id,
+            'receiver_id' => Checkout::studentsCouncil()->handler_id,
```
Comment (kdmnk) [inline app/Models/User.php]: the receiver can be any collegist that the student council delegates to, not necessarily the handler. In both cases.

Comment (kdmnk) [reply app/Models/User.php]: Despite how they do this nowadays.

File: app/Models/User.php [hunk 5]
```diff
@@ -869,25 +863,64 @@ public function sumOfActiveFreePages(): int
...
+    }
+
+    /**
+     * Returns the payed kkt amount in the semester; or null if the user has not payed kkt.
      * @param Semester $semester
-     * @return int
+     * @return ?int
      */
-    public function payedKKTInSemester(Semester $semester): int
+    public function payedKKTInSemester(Semester $semester): ?int
```
Comment (kdmnk) [inline app/Models/User.php]: Would you rename everything to "paidSomething" while you are here, please?

File: app/Models/Workshop.php [hunk 6]
```diff
@@ -69,6 +70,23 @@ public function externs()
...
+        return $this->hasMany(WorkshopBalance::class);
+    }
+
+    /**
+     * Returns the balance for a given semester
+     * (by default the current one).
+     */
+    public function balance(int $semester = null): WorkshopBalance
+    {
+        return $this->balances()->where('semester_id', $semester ?? Semester::current()->id)->firstOrFail();
```
Comment (kdmnk) [inline app/Models/Workshop.php]: Won't this die if we switch to next semester and the current workshop balances are not generated yet?
I'd allow null value and handle that.

File: app/Models/WorkshopBalance.php [hunk 7]
```diff
@@ -58,7 +65,7 @@ public static function generateBalances($semester_id)
             $not_yet_paid = 0;
             foreach ($workshop->users as $member) {
                 if (isset($active_users[$member->id])) {
-                    if (!isset($users_has_to_pay_kktnetreg[$member->id])) {
+                    if (!is_null($member->payedKKTInSemester(Semester::find($semester_id)))) {
```
Comment (kdmnk) [inline app/Models/WorkshopBalance.php]: Get the semester outside of the loop. Or even outside this function and get the semester model as param. It makes our type checking stronger.

File: tests/Unit/EconomicTest.php [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+    /**
+     * Tests how workshop balances change
+     * when a user pays kkt.
+     * Does it both for externs and residents.
+     */
+    public function testPayment()
+    {
+        foreach ([true, false] as $is_extern) {
```
Comment (kdmnk) [inline tests/Unit/EconomicTest.php]: nice trick, but maybe cleaner to put these into two separate test cases.

File: tests/Unit/EconomicTest.php [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+            $workshop_id = rand(1, count(Workshop::ALL));
+            $workshop = Workshop::findOrFail($workshop_id);
```
Comment (kdmnk) [inline tests/Unit/EconomicTest.php]: You can use Workshop::all()->random() and then get the id from that.

File: tests/Unit/EconomicTest.php [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+            $user = User::factory()->create();
+
+            if ($is_extern) {
+                $user->setExtern();
+            } else {
+                $user->setCollegist(\App\Models\Role::RESIDENT);
```
Comment (kdmnk) [inline tests/Unit/EconomicTest.php]: ->setResident?

File: tests/Unit/EconomicTest.php [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+            }
+
+            $user->setStatusFor(Semester::current(), SemesterStatus::ACTIVE);  // important!
+
+            $user->workshops()->attach($workshop_id);
```
Comment (kdmnk) [inline tests/Unit/EconomicTest.php]: FYI I believe you can simply pass the workshop, but whatever

File: tests/Unit/EconomicTest.php [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+
+            $with_same_status =
+                $is_extern ? $workshop->externs() : $workshop->residents();
+            $this->assertTrue(
+                $with_same_status->filter(function (User $u) use ($user) {return $u->id == $user->id;})->count()
+                  == 1
+            );
```
Comment (kdmnk) [inline tests/Unit/EconomicTest.php]: What's the point here? this only checks that you set previously and that the user does not appear twice in the collection which can never be the case, can it?

File: tests/Unit/EconomicTest.php [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+            $user->payKKTNetreg(config('custom.kkt'), config('custom.netreg'));
+            $balance->refresh();
+            $new = $balance->allocated_balance;
+
+            $this->assertTrue($user->payedKKT() == config('custom.kkt'));
+
+            $this->assertTrue($new - $old == round(
+                ($is_extern ? config('custom.workshop_balance_extern')
+                                                           : config('custom.workshop_balance_resident'))
+                                               * config('custom.kkt')
```
Comment (kdmnk) [inline tests/Unit/EconomicTest.php]: Use precoded values instead of config values to make this easier to read and debug. You can have another test case that the pay function uses the config values accordingly.

**Review by kdmnk — CHANGES_REQUESTED** [review]
Okay looking better. Please take a look at my comments. And yes, please write a case to test multiple workshops, that's a common case we have. So I imagine 4 cases, combining resident/extern and one/multiple workshops. 
Have you found the original bug yet?

An HTTP test is not necessarily here as we test those anyways during development. Those are useful for refactorings though, but we most importantly lack unit tests to test different edge cases like these.

File: tests/Unit/EconomicTest.php (hunk 12 above)
Comment (viktorcsimma) [reply tests/Unit/EconomicTest.php]: The point was that I wanted to know that the status setting occurred successfully. But I agree it does not belong here.

File: app/Models/User.php (hunk 4 above)
Comment (viktorcsimma) [reply app/Models/User.php]: I now added it as a parameter.

File: app/Models/User.php (hunk 5 above)
Comment (viktorcsimma) [reply app/Models/User.php]: I don't understand this here; should I generalise the function to be able to pay anything? Now it is specially about kkt.

File: app/Models/User.php (hunk 3 above)
Comment (viktorcsimma) [reply app/Models/User.php]: Oh; it was the cache:D After running `php artisan optimize:clear`, it was fine. I revert it.

File: tests/Unit/EconomicTest.php (hunk 13 above)
Comment (viktorcsimma) [reply tests/Unit/EconomicTest.php]: I had a problem with the ratios, or rather a design question: now `generateBalances` itself uses the config values for them; should we provide them as parameters or store them as properties of the semester? Or simply leave this as it is?

**Comment by viktorcsimma** [comment]
Now I've added new test cases; I've also removed the first one (there was some problem with it, and you have mentioned it's not useful anyway).

Can you check out these and my questions, too? Sometime when it's convenient for you – until then, happy holidays;)

File: app/Http/Controllers/StudentsCouncil/EconomicController.php [hunk 14]
```diff
@@ -85,41 +85,22 @@ public function payKKTNetreg(Request $request)
         $validator->validate();
 
         $payer = User::findOrFail($request->user_id);
+        // the current user will be the receiver
+        [$kkt, $netreg, $new_internet_expire_date] = $payer->payKKTNetreg(Auth::user()->id, $request->kkt, $request->netreg);
```
Comment (kdmnk) [inline app/Http/Controllers/StudentsCouncil/EconomicController.php]: As a new thought, this adds another method to the User model which we want to keep as simple as possible.
I was thinking a lot about where would this be placed the best, in the Transaction model or somehow in CheckoutHandler, but these actions are typical controller actions, so you don't need to extract them away. We can make the function smaller with automatic email sending with observers, but that would mean refactoring the whole CheckoutHandler.
Just revert this change for now, sorry :D

File: app/Models/User.php (hunk 5 above)
Comment (kdmnk) [reply app/Models/User.php]: Just rename it to make it gramatically correct, paidKKT and paidNetreg.

File: tests/Unit/EconomicTest.php [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+        } else {
+            $user->setResident();
+        }
+
+        $user->setStatusFor(Semester::current(), SemesterStatus::ACTIVE);  // important!
+        foreach($workshops as $workshop) {
+            $user->workshops()->attach($workshop);
+        }
+
+        $this->assertTrue(is_null($user->payedKKT()));
```
Comment (kdmnk) [inline tests/Unit/EconomicTest.php]: unnecessary

File: tests/Unit/EconomicTest.php [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+        WorkshopBalance::generateBalances(Semester::current()); // this also ensures the balance won't be null
+
+        $balances = array_map(function ($workshop) {return $workshop->balance();}, $workshops);
```
Comment (kdmnk) [inline tests/Unit/EconomicTest.php]: use collection methods. 
$old_balances = $workshops->map(fn ($workshop) => $workshop->balance()->allocated_balance). 
$new_balances = same as above, i don't even think you need to refresh the vars this way.

File: tests/Unit/EconomicTest.php [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+        // we give the payer's id as the receiver's id
+        // because Checkout::studentsCouncil()->handler_id returns null
+        $user->payKKTNetreg($user->id, self::TEST_KKT, self::TEST_NETREG);
+
+        $this->assertTrue($user->payedKKT() == self::TEST_KKT);
+
+        $news = array_map(function ($balance) {return $balance->fresh()->allocated_balance;}, $balances);
+
+        $amount_for_workshops =
+            ($is_extern ? config('custom.workshop_balance_extern')
```
Comment (kdmnk) [inline tests/Unit/EconomicTest.php]: remove these configs as well. This also ensures that the pipelines test with the same data.

File: tests/Unit/EconomicTest.php [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+                        : config('custom.workshop_balance_resident'))
+                * self::TEST_KKT;
+        for ($i = 0; $i < $number_of_workshops; ++$i) {
+            // to avoid floating point equality checking
+            // and to allow some error from rounding:
+            $this->assertTrue(abs(
+                $number_of_workshops * ($news[$i] - $olds[$i])
+                                - $amount_for_workshops
+            ) < self::THRESHOLD_DIFF);
+        }
```
Comment (kdmnk) [inline tests/Unit/EconomicTest.php]: This is not readable. Use assertEqualsWithDelta and avoid for-loop indexes. Loop through the workshops, use dicts or whatever. Even easier to check if the workshop balances equal to an expected hard-coded value that you calculated before, no need to calculate the diffs programatically. Trust me it's better for debugging tests.

You should be able to set the threshold below 0.1. If not, something's wrong. If we round, still, the sum should all add up to the originally paid amount. Maybe you need to modify the logic behind it.

File: tests/Unit/EconomicTest.php [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+        $workshops = Workshop::all()->random($number_of_workshops)->all();
+        $user = User::factory()->create();
+
+        if ($is_extern) {
+            $user->setExtern();
+        } else {
+            $user->setResident();
+        }
+
+        $user->setStatusFor(Semester::current(), SemesterStatus::ACTIVE);  // important!
```
Comment (kdmnk) [inline tests/Unit/EconomicTest.php]: ->setStatus(SemesterStatus::ACTIVE);

File: tests/Unit/EconomicTest.php [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+        foreach($workshops as $workshop) {
+            $user->workshops()->attach($workshop);
+        }
```
Comment (kdmnk) [inline tests/Unit/EconomicTest.php]: $user->workshops()->sync(Workshop::all()->random(2))

File: tests/Unit/EconomicTest.php [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+    private function testPayment(bool $is_extern, int $number_of_workshops = 1): void
+    {
+        $workshops = Workshop::all()->random($number_of_workshops)->all();
+        $user = User::factory()->create();
+
+        if ($is_extern) {
+            $user->setExtern();
+        } else {
+            $user->setResident();
+        }
```
Comment (kdmnk) [inline tests/Unit/EconomicTest.php]: $user->setCollegist($is_extern ? Role::EXTERN : Role::RESIDENT)

File: tests/Unit/EconomicTest.php [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+    {
+        $this->testPayment(false);
+    }
+
+    /**
+     * The concrete call of testPayment for externs with two workshops.
+     */
+    public function testExternPaymentWithTwoWorkshops()
+    {
+        $this->testPayment(true, 2);
```
Comment (kdmnk) [inline tests/Unit/EconomicTest.php]: These are also not readable. How about passing the user as a parameter?
I suggested some changes to make the user creation logic smaller above. 

I might also move the assertion logic here with the hard-coded amounts described above.
And with that, you probably won't even need these functions :D

**Review by kdmnk — CHANGES_REQUESTED** [review]
These comments are just to make the code more readable. Good job

**Comment by kdmnk** [comment]
I'm not saying that it's perfect, but you can take a look at these tests, here the assertions are clear what they tests and how they got calculated (eg. hard coded  `400 * (100 - (100 - 25) / 2)`):
https://github.com/DevsWithDodo/dodo_api/blob/master/tests/Feature/PurchaseCreateTest.php

File: app/Models/User.php (hunk 5 above)
Comment (viktorcsimma) [reply app/Models/User.php]: Oh, I see:)

File: app/Http/Controllers/StudentsCouncil/EconomicController.php (hunk 14 above)
Comment (viktorcsimma) [reply app/Http/Controllers/StudentsCouncil/EconomicController.php]: What about writing a static method in the EconomicController class? That was where this code was originally written. I would rather write a method for it somewhere, so that we don't need to duplicate the code in the test. See the code I'm going to push soon.

File: tests/Unit/EconomicTest.php (hunk 22 above)
Comment (viktorcsimma) [reply tests/Unit/EconomicTest.php]: Passing the user is a good idea; that way, it's more evident what we are doing while we also avoid much of code duplication. I think I'll go for that:)

**Review by kdmnk — APPROVED** [review]

**Comment by kdmnk** [comment]
Sorry about the hassle of such a little feature, but I think it looks really good and clean now. Merge as you wish

**Comment by viktorcsimma** [comment]
It's fine; thank you for your help:) eventually, I'll have to know how to sustain this code quality.
